tx-pool/src/service.rs

```diff
@@ -400,11 +402,48 @@ impl TxPoolController {
             .map_err(handle_recv_error)
             .map_err(Into::into)
     }
+
+    /// Saves tx pool into disk.
+    pub fn save_pool(&self) -> Result<(), AnyError> {
+        ckb_logger::info!("Please be patient, TxPool are saving data into disk ...");
+        let mut sender = self.sender.clone();
+        let (responder, response) = oneshot::channel();
+        let request = Request::call((), responder);
+        sender.try_send(Message::SavePool(request)).map_err(|e| {
+            let (_m, e) = handle_try_send_error(e);
+            e
+        })?;
+        self.handle.block_on(response).map_err(handle_recv_error)?
+    }
+
+    fn load_persisted_data(&self, data: persisted::TxPool) -> Result<(), AnyError> {
+        // a trick to commit transactions with the correct order
+        let mut remain_size = data.transactions().len();
+        let mut txs_next_turn = Vec::new();
+        for tx in data.transactions() {
+            let tx_view = tx.into_view();
+            if self.submit_local_tx(tx_view.clone())?.is_err() {
+                txs_next_turn.push(tx_view);
+            }
+        }
+        while !txs_next_turn.is_empty() && remain_size != txs_next_turn.len() {
```

Comment:
There is a potential dead loop here: because chain and pool processing is asynchronous, if the ckb is closed between the time of chain inserts a block and removes the committed txs from pool, the dumped transactions will never replay successfully, resulting in a dead loop. Suggest to add a loop count limit to replay txs.

Reply:
> There is a potential dead loop here: because chain and pool processing is asynchronous, if the ckb is closed between the time of chain inserts a block and removes the committed txs from pool, the dumped transactions will never replay successfully, resulting in a dead loop. Suggest to add a loop count limit to replay txs.

`remain_size != txs_next_turn.len()`

If in any turn there are no transaction was processed, then this loop will be broken.

Reply:
> because chain and pool processing is asynchronous

Yes, I spent some hours on that.
Before tx-pool rewrite, I could drop network and rpc before tx-pool, also I could create tx-pool before network and rpc.

I really don‘t like current implementation, too.
I except that I could shutdown all other components before save tx-pool, and load tx-pool before start any other components.
But I didn't find a way to do that.

Reply:
aha, I didn't notice the `remain_size != txs_next_turn.len()` code here, however, is it easier to read by using `loop break`?
```
        let mut txs = data.transactions().into_iter().map(|tx| tx.into_view()).collect::<Vec<_>>();
        loop {
            let mut failed_txs = Vec::new();
            for tx in txs.iter() {
                if self.submit_local_tx(tx.clone())?.is_err() {
                    failed_txs.push(tx.clone());
                }
            }
            // return when all success or all failed
            if failed_txs.is_empty() || txs.len() == failed_txs.len() {
                break;
            }
            txs = failed_txs;
        }
        Ok(())
```